File: content/snippets/cpp/interview_lcci/01/07_rotate_matrix.cpp
#include <iostream>
#include <vector>

void rotate(std::vector<std::vector<int>>& m) {
    const int n = static_cast<int>(m.size());
    for (int layer = 0; layer < n / 2; ++layer) {
        int first = layer;
        int last = n - 1 - layer;
        for (int i = first; i < last; ++i) {
            int offset = i - first;
            int top = m[first][i];
            m[first][i] = m[last - offset][first];
            m[last - offset][first] = m[last][last - offset];
            m[last][last - offset] = m[i][last];
            m[i][last] = top;
        }
    }
}

int main() {
    std::vector<std::vector<int>> m = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    };

    rotate(m);

    for (const auto& row : m) {
        std::cout << row[0] << ' ' << row[1] << ' ' << row[2] << '\n';
    }
    return 0;
}

File: content/snippets/python/interview_lcci/01/07_rotate_matrix.py
def rotate(matrix: list[list[int]]) -> None:
    n = len(matrix)
    for layer in range(n // 2):
        first = layer
        last = n - 1 - layer
        for i in range(first, last):
            offset = i - first
            top = matrix[first][i]
            matrix[first][i] = matrix[last - offset][first]
            matrix[last - offset][first] = matrix[last][last - offset]
            matrix[last][last - offset] = matrix[i][last]
            matrix[i][last] = top


def main() -> None:
    matrix = [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9],
    ]

    rotate(matrix)
    for row in matrix:
        print(*row)


if __name__ == "__main__":
    main()
